Manage extended table attributes

eusdougc opened the pull request:
This pull request opens the possibility of importing more complicated open source models in a variety of formats.  Previously, the import action created the attributes for an image classification model.  Now if the dataSpec parameter is supplied, more complicated scenarios (e.g. object detection) are supported.  The changes will be both forward and backward compatible because I test whether the import action supports the dataSpec parameter.  If it doesn't, I create the model attributes using Python code.  If it does, I let the action create the correct attributes.

I ran all the unit tests and there were no errors.

ethem-kinginthenorth commented on dlpy/network.py:
```diff
@@ -752,18 +817,72 @@ def load_weights_from_file_with_labels(self, path, format_type='KERAS'):
+            
+        if (data_spec):
+            # see if dataspecs supported by action
+            from io import StringIO
```
for StringIO we might need seomthing like this:
# Exception handler for reading StringIO text data into memory
try:
    from StringIO import StringIO
except ImportError:
    from io import StringIO

eusdougc replied:
I can add the try/except block as you suggest.

ethem-kinginthenorth commented on dlpy/utils.py:
```diff
@@ -387,6 +387,20 @@ def get_imagenet_labels_table(conn):
     return conn.CASTable(temp_name)
 
 
+def get_user_defined_labels_table(conn, label_file_name):
```
can you also change the get_imagenet_labels_table function to call this one.

eusdougc replied:
Yes - I assume the intention here is to have something like this:

def get_imagenet_labels_table(conn):
    temp_name = random_name('new_label_table', 6)

    filename = os.path.join('datasources', 'imagenet_labels.csv')
    project_path = os.path.dirname(os.path.abspath(__file__))
    full_filename = os.path.join(project_path, filename)
    
    return (get_user_defined_labels_table(conn, full_filename))

kesmit13 commented on dlpy/attribute_utils.py:
```diff
@@ -0,0 +1,763 @@
+        raise DLPyError('Cannot export model attributes, there seems to be a problem.')
+
+    ascii_text = rt['xmlblob'].decode('utf8')
+    myfile = open(file_name, "w")  
```
You might want to use a context manager here: 
```
with open(file_name, "w") as myfile:
    myfile.write(ascii_text)
```

eusdougc replied:
I can add this.

kesmit13 commented on dlpy/network.py:
```diff
@@ -726,22 +745,68 @@ def load_weights_from_file(self, path, format_type='KERAS'):
-                                          name=self.model_name + '_weights'),
-                        formatType=format_type, weightFilePath=file_name,
-                        caslib=cas_lib_name)
+        if (data_spec):
```
BTW, Python doesn't need parens around 'if' conditions.  I see that in various places in this code.

eusdougc replied:
Agreed, but I find it helps readability (at least for me).  If this is a style problem, I can certainly make the change, but otherwise would prefer to leave it as it is.

kesmit13 commented on dlpy/network.py:
```diff
@@ -726,22 +745,68 @@ def load_weights_from_file(self, path, format_type='KERAS'):
+            import sys
+
+            # redirect stderr
+            old_stderr = sys.stderr
```
I'm not sure why this remapping of stderr is here.  That's kind of a sketchy thing to do and I'm not sure what it is accomplishing here.  The message level of the action is set to none, so it shouldn't be sending anything back as it is.

eusdougc replied:
I didn't like doing this but it was the only way to suppress the "ERROR: ..." returned by the action when the dataSpec parameter was not defined.  Despite setting the messagelevel to NONE, the error was always returned.  I tried a number of things to suppress the error, including setting the CAS session option for suppressing messages and nothing worked.

kesmit13 replied:
It may be that there's a bug in the action so that it's still sending the error messages back.  However, I would think that using a context manager like the following would prevent anything from being printed:
```
with swat.option_context(print_messages=False):
    ... code with messages to suppress ...
```

eusdougc replied:
Ethem just walked over and made the same suggestion.  I tested it and it seems to work so I'm replacing the redirection of stderr with the context manager.

kesmit13 commented on dlpy/network.py:
```diff
@@ -726,22 +745,68 @@ def load_weights_from_file(self, path, format_type='KERAS'):
+                                modelWeights=dict(replace=True, name=self.model_name + '_weights'),
+                                dataSpecs=data_spec,
+                                formatType=format_type, weightFilePath=file_name, caslib=cas_lib_name,
+                                #_debug = dict(nodes="all",ranks='-1',roles="all",display="d7a752.na.sas.com:0")
```
You should remove this debug parameter.  It contains a SAS hostname.

eusdougc replied:
Thanks, I missed this when I did my push.

kesmit13 commented on dlpy/network.py:
```diff
@@ -726,22 +745,68 @@ def load_weights_from_file(self, path, format_type='KERAS'):
+                                );
+            
+            # restore stderr
+            sys.stderr = old_stderr                
```
Same as above.

eusdougc replied:
Please see my reply to your first comment on this topic.

kesmit13 commented on dlpy/network.py:
```diff
@@ -726,22 +745,68 @@ def load_weights_from_file(self, path, format_type='KERAS'):
+                                                      name=self.model_name + '_weights'),
+                                    formatType=format_type, weightFilePath=file_name,
+                                    caslib=cas_lib_name, 
+                                    #_debug = dict(nodes="all",ranks='-1',roles="all",display="d7a752.na.sas.com:0")
```
Same as above.

eusdougc replied:
I will remove this.

kesmit13 commented on dlpy/network.py:
```diff
@@ -726,22 +745,68 @@ def load_weights_from_file(self, path, format_type='KERAS'):
+                                              name=self.model_name + '_weights'),
+                            formatType=format_type, weightFilePath=file_name,
+                            caslib=cas_lib_name, 
+                            #_debug = dict(nodes="all",ranks='-1',roles="all",display="d7a752.na.sas.com:0")
```
Same as above.

eusdougc replied:
I will remove this.

kesmit13 commented on dlpy/network.py:
```diff
@@ -752,18 +817,72 @@ def load_weights_from_file_with_labels(self, path, format_type='KERAS'):
+            from io import StringIO
+            import sys
+
+            # redirect stderr
```
Same as above for this whole block of code.

eusdougc replied:
Same response as to your original two comments on this topic.  If there is another more CAS-friendly way to suppress the error messages, I'd prefer to do it that way.

eusdougc commented on dlpy/attribute_utils.py:
```diff
@@ -0,0 +1,763 @@
+
+    Parameters
+    ----------
+    conn : CAS
```
If that is true, then the headers for many functions will need to be changed throughout DLPy.  I just copied the header from a function in utils.py since I was using the conn object in exactly the same way.  I just did a quick check to verify that this object is used similarly in other files and has exactly the same description throughout.

eusdougc commented:
Just pushed new versions of utils.py, network.py, and attribute_utils.py.  These address the following issues:

1) Python if style (i.e. no parenthesis around condition)
2) Removed stderr redirection in favor of a context manager
3) Updated get_imagenet_label_tables function to use get_user_defined_labels_table

eusdougc commented:
I did run the existing unit tests and there were no errors.  I do not have a unit test for this feature as yet.  I assume you want any unit test to run quickly, so I cannot use the ONNX import scenario that I've been using to validate my approach.  I believe I can build something smaller, but I will have to look over the existing unit tests to construct an appropriate test.
